scripts/controllers/RobotArm.py

```diff
@@ -0,0 +1,198 @@
+#!/usr/bin/env python
+"""
+Code.
+"""
+import rospy
+import numpy as np
+from std_msgs.msg import Float64
+from sensor_msgs.msg import JointState
+from trajectory_msgs.msg import JointTrajectory, JointTrajectoryPoint
+from RobotControllerManager import RobotControllerManager, ControllerType
+
+
+class RobotArm(object):
+    """
+    Generic robot arm implementation in ROS. Works for n joints.
+    """
+
+    def __init__(self, num_joints,
+                 is_sim=True, joint_states_topic='/joint_states',
+                 ignore_joints_arr=['panda_finger_joint1', 'panda_finger_joint2'],
+                 controller_names=None):
+        """
+        intializiation of the robot arm class.
+
+        `controller_names` should have:
+
+        `controller_names[0]` - position controllers
+
+        `controller_names[1]` - velocity controllers
+
+        `controllers_names[2]` - ONE trajectory controller.
+        """
+
+        self.num_joints = num_joints
+        if controller_names is None:
+            controller_names = []
+            default_pos_names = ['panda_joint{}_position_controller'.format(i) for i in range(1, num_joints+1)]
+            default_vel_names = ['panda_joint{}_velocity_controller'.format(i) for i in range(1, num_joints+1)]
+            default_traj_name = ['panda_joint_trajectory_controller' if is_sim else 'joint_trajectory_controller']
+            controller_names.append(default_pos_names)
+            controller_names.append(default_vel_names)
+            controller_names.append(default_traj_name)
+
+        # create the controller manager for switching controllers.
+        self.controller_manager = RobotControllerManager(controller_names)
+
+        self.ignore_joints_arr = ignore_joints_arr
+        self.data_exists = False
+        self.joint_data = None
+        self.names = None
+        self.positions = None
+
+        # get list of position and velocity publishers ... and trajectory publisher.
+        self.pos_pubs, self.vel_pubs, self.traj_pub = self.get_publishers(controller_names)
+
+        rospy.Subscriber(joint_states_topic, JointState, self.joint_state_callback)
+        # need a sleep, so message is sent from publisher, interestingly.
+        rospy.sleep(2)
+
+    def joint_names(self):
+        """
+        get the joint names of the robot
+        """
+        if self.data_exists:
+            return self.names
```

Comment:
Does this mean that nothing will be returned if `RobotArm` did not receive any data from joint publishers?

Reply:
yea.. I think I will just return an empty list there.

Reply:
done